### ISSUE TITLE
[ENH] Make it possible to include `sample_weight` when using model_evaluate `evaluate`

### ISSUE THREAD

**Issue by MarkusSagen** [sktime/sktime#7279]
**Is your feature request related to a problem? Please describe.**
Using the `evaluate` method from sktime's model_evaluation, it's currently not possible to provide custom sample weights

**Describe the solution you'd like**
Most performance metrics can use sample weights. Since evaluate takes in a splitter strategy (cv), sample weights need to to align with the splits. There are several potential approaches to this

1. Allow callbacks in the evaluate method. This approach leaves future extensions of the method to the users. In the future, it is likely that others may have requests to add other features to the evaluate method. Instead of providing multiple arguments, the method is extensible for others needs
2. Include a new argument to the evaluate method with a sample_weight method or strategy.

**Describe alternatives you've considered**

**Additional context**
One approach I've already tested that works is to modify these lines:
https://github.com/sktime/sktime/blob/974ceface95b3ae159596525550be5d02cacfc93/sktime/forecasting/model_evaluation/_functions.py#L261-L262

```diff
- score = metric(y_test, y_pred, y_train=y_train)

+ sample_weights = sample_weight_strategy(y_test, **sample_weight_params)
+ score = metric(y_test, y_pred, y_train=y_train)
```

**Comment by fkiraly** [comment]
Interesting! Thoughts and questions:

* where does `sample_weight_strategy` come from? What would be complete example usage?
* very much related: https://github.com/sktime/sktime/issues/7276 - this is another example for an argument to a method that cannot be passed in `evaluate`, the difference is that it needs to be passed to `predict_interval` rather than the metric directly. I wonder whether there is a common approach here.

**Comment by fkiraly** [comment]
Hm, interesting. What would be the signature/type of `sample_weight_strategy` - would it be a function?

**Comment by MarkusSagen** [comment]
I played around a bit with it and this implementation seems to work.
Essentially, it would create a new base class and extend BaseForecastingErrorMetric with another argument: `sample_weight_generator`

Sample weights would still be passed in as a kwargs argument to work with other metrics and classes


Base
```py
import typing as T
from abc import ABC, abstractmethod
from inspect import getfullargspec, isfunction, signature
from typing import Protocol, runtime_checkable, TYPE_CHECKING

if TYPE_CHECKING:
    import numpy as np


@runtime_checkable
class SampleWeightGenerator(Protocol):
    def __call__(self, y_true, y_pred=None, **kwargs) -> np.ndarray | None: ...


class BaseSampleWeightGenerator(ABC):
    @abstractmethod
    def __call__(self, y_true, y_pred=None, **kwargs) -> np.ndarray | None:
        """Generate sample weights for a given set of true and predicted values.

        Parameters
        ----------
        y_true : time series in sktime compatible data container format
            Ground truth (correct) target values.

        y_pred : time series in sktime compatible data container format, optional
            Predicted values to evaluate against ground truth.
            Must be of same format as y_true, same indices and columns if indexed.
            If None, only y_true is used for weight generation.

        **kwargs : dict
            Additional keyword arguments specific to the weight generation method.

        Returns
        -------
        sample_weight : 1D np.ndarray or None
            Sample weights for each time point.

            * If an array, must be 1D.
              If y_true and y_pred are a single time series,
              sample_weight must be of the same length as y_true.
              If the time series are panel or hierarchical, the length of all
              individual time series must be the same, and equal to the length
              of sample_weight, for all instances of time series passed.
            * If None, the time indices are considered equally weighted.
        """
        ...
```

**Comment by MarkusSagen** [comment]
Then we could implement custom sample weight generators 

```py
class DummyWeightGenerator(BaseSampleWeightGenerator):
    def __call__(self, y_true: np.ndarray, y_pred: np.ndarray | None = None, **kwargs) -> None:
        return None


class UniformWeightGenerator(BaseSampleWeightGenerator):
    def __call__(self, y_true: np.ndarray, y_pred: np.ndarray | None = None, **kwargs) -> np.ndarray:
        return np.ones_like(y_true)


class RecencyWeightGenerator(BaseSampleWeightGenerator):
    def __init__(self, decay_rate: float = 0.1):
        self.decay_rate = decay_rate

    def __call__(self, y_true: np.ndarray, y_pred: np.ndarray | None = None, **kwargs) -> np.ndarray:
        weights = np.exp(-self.decay_rate * np.arange(len(y_true))[::-1])
        return weights / np.sum(weights)


class DateRangeWeightGenerator(BaseSampleWeightGenerator):
    def __init__(self, date_ranges: list[dict], default_weight: float = 1.0):
        self.date_ranges = date_ranges
        self.default_weight = default_weight

    def __call__(self, y_true: np.ndarray, y_pred: np.ndarray | None = None, **kwargs) -> np.ndarray:
        if "dates" not in kwargs:
            raise ValueError("'dates' must be provided in kwargs for DateRangeWeightGenerator")

        dates = pd.to_datetime(kwargs["dates"])
        weights = np.full(len(y_true), self.default_weight)

        for range_config in self.date_ranges:
            start_date = pd.to_datetime(range_config["start"], format="%m-%d")
            end_date = pd.to_datetime(range_config["end"], format="%m-%d")
            mask = (
                (dates.dt.month >= start_date.month)
                & (dates.dt.month <= end_date.month)
                & (dates.dt.day >= start_date.day)
                & (dates.dt.day <= end_date.day)
            )
            weights[mask] = range_config["weight"]

        return weights / np.sum(weights)


class SeasonalWeightGenerator(BaseSampleWeightGenerator):
    def __init__(self, winter_weight=8, summer_weight=4, default_weight=1):
        self.winter_weight = winter_weight
        self.summer_weight = summer_weight
        self.default_weight = default_weight

    def __call__(self, y_true: np.ndarray, y_pred: np.ndarray | None = None, **kwargs) -> np.ndarray:
        if "dates" not in kwargs:
            raise ValueError("'dates' must be provided in kwargs for SeasonalWeightGenerator")

        dates = pd.to_datetime(kwargs["dates"])
        weights = np.full(len(y_true), self.default_weight)

        # Winter weights (November to February)
        winter_mask = (dates.dt.month >= 11) | (dates.dt.month <= 2)
        weights[winter_mask] = self.winter_weight

        # Summer weights (June to August)
        summer_mask = (dates.dt.month >= 6) & (dates.dt.month <= 8)
        weights[summer_mask] = self.summer_weight

        return weights / np.sum(weights)
```

**Comment by MarkusSagen** [comment]
And call then in a complete example like this:

```py
import numpy as np
import pandas as pd
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.naive import NaiveForecaster

np.random.seed(42)

dates = pd.date_range(start="2020-01-01", end="2022-12-31", freq="D")

t = np.arange(len(dates))
y = 10 + 0.05 * t + 5 * np.sin(2 * np.pi * t / 365) + np.random.normal(0, 1, len(t))
df = pd.DataFrame({"ds": dates, "y": y})

train_df, test_df = temporal_train_test_split(df, test_size=0.2)

# Convert the data to the format expected by sktime
y_train = pd.Series(train_df["y"].values, index=train_df["ds"])

fh = np.arange(1, len(test_df) + 1)

model = NaiveForecaster(strategy="last")
model.fit(y_train)

y_pred = model.predict(fh)
y_pred = y_pred.to_numpy()
y_true = test_df["y"].to_numpy()

# Create the weight generator
weight_generator = SeasonalWeightGenerator(winter_weight=8, summer_weight=4, default_weight=1)

# Create the metric with the custom weight generator
msse = MeanSquaredScaledError()
wmsse = MeanSquaredScaledError(sample_weight_generator=weight_generator)

# Calculate the weighted MSSE
unweighted_msse = msse(y_true, y_pred, y_train=train_df["y"].values)
weighted_msse = wmsse(y_true, y_pred, y_train=train_df["y"].values, dates=test_df["ds"])

print(f"Weighted MSSE: {weighted_msse:.4f}")
print(f"Unweighted MSSE: {unweighted_msse:.4f}")
```

**Comment by fkiraly** [comment]
@MarkusSagen, thanks!

In chess, this is the point where a spectator could now write `!` or `!!`.
https://en.wikipedia.org/wiki/Chess_annotation_symbols#!_(Good_move)

It does seem like you have a fork of `sktime` somewhere with this working completely in integration - would you be so kind and open a PR?

I would tend towards being in favour and merging, but we need to deal with things such as tests, file locations, core developer review.

It seems fully downwards compatible, since we add only a new argument and default is as before.


Major esign comment:

Instead of adding a new algorithm, can we not just use the `sample_weight` argument, which now can be a function in addition to iterables? That would avoid a major change to the base metric interface.

However, I would suggest you make a draft PR with whatveer you have, and we use that as a basis for discussion.

**Comment by MarkusSagen** [comment]
Good point! I can open a PR for this

I think though, if we change the kwarg of sample_weight, many metrics reference sample weights as being an iterable (numpy values) or none. Changing the behavior of it would mean we also should update documentation for it?

A reason for proposing one additional argument in BaseForecastingErrorMetric is that closed methods, s.a. the method `evaluate`; that makes it hard to include sample weights in other methods where they are called.

Would you say it would be a better approach maybe to extend the __init__ for BaseForecastingErrorMetric to accept kwargs instead or something else?

I would think it is a bit easier to do it this way, including one additional argument, even if that would require adding some additional documentation to metrics?

I can open a draft PR for this

**Comment by fkiraly** [comment]
> many metrics reference sample weights as being an iterable (numpy values) or none. Changing the behavior of it would mean we also should update documentation for it?

Yes, but there is a single point of change here:

* for docstring, the docstrings for `evaluate` and `evaluate_by_index` in the base class
* for the logic to get weights (iterable/array) from the function, the boilerplate in the same functions

Therefore, we need to change nothing in the concrete metrics if we use the existing argument `sample_weights`.

Regarding accepting `kwargs`, that is problematic, as it clashes with the `sklearn`-like API, where arguments are expected to be all explicit.

**Comment by MarkusSagen** [comment]
Okey, I'll take a stab at it and make a draft re-working sample weights to be a callable or weights in the __call__ method instead

**Comment by MarkusSagen** [comment]
Would consider this closed and done
This link shows how we can implement custom sample weights now too

https://github.com/sktime/sktime/commit/37d601786c27fbaed084dc6b79194c6d5554f648
